GPII-1280: Initial implementation of user management...

the-t-in-rtf opened the pull request:
See [GPII-1280](https://issues.gpii.net/browse/GPII-1280) for details.

amb26 commented on .jshintrc:
```diff
@@ -0,0 +1,32 @@
+
+    "node": true,
+
+    "globals": {
```
delete

the-t-in-rtf replied:
Done.

amb26 commented on src/js/client/canHandleStrings.js:
```diff
@@ -0,0 +1,57 @@
+    fluid.registerNamespace("gpii.express.user.frontend.canHandleStrings");
+
+    gpii.express.user.frontend.canHandleStrings.stringToJSON = function (that, data) {
+        if (typeof data === "string") {
```
As seen in previous contexts, this "convert to JSON if you can but otherwise do nothing" is a suspicious architectural smell - it implies that the workflow is not clear on what type of data it is expecting, and will lead to problems in the future.

the-t-in-rtf replied:
I suspect this grade was covering for earlier crappiness in bad AJAX responses either from the previous user management library, from my pre `gpii-express` API code, or from both.

I was able to remove this grade entirely and all tests pass, so all client-side components in this package now accept only JSON data in responses and fail if they don't receive it.

amb26 commented on src/js/client/controls.js:
```diff
@@ -0,0 +1,150 @@
+    var gpii = fluid.registerNamespace("gpii");
+    fluid.registerNamespace("gpii.express.user.frontend.controls");
+
+    gpii.express.user.frontend.controls.handleMenuKeys = function (that, event) {
```
Consider the use of the Infusion Keyboard Accessibility plugin "activatable" in order to collect this information in a more easily visible declarative form: 
https://wiki.fluidproject.org/display/docs/Keyboard+Accessibility+Plugin+API
(docs are old, but still current).
We have an old tutorial here:
https://wiki.fluidproject.org/display/Infusion14/Tutorial+-+Keyboard+Accessibility+Plugin
Ideally, the arguments to the activatable plugin will just be captured in a set of mixin grade definition options.

amb26 replied:
Up to date demo sample: https://github.com/fluid-project/infusion/blob/master/demos/keyboard-a11y/js/five-star.js#L121

the-t-in-rtf replied:
I was able to update the component to use both `activatable` and `selectable` and removed most of my unique event handling bits in favor of those standard approaches.

amb26 commented on src/js/client/controls.js:
```diff
@@ -0,0 +1,150 @@
+            }
+        },
+        listeners: {
+            onMarkupRendered: [
```
Supply these with namespaces rather than sticking in a large array so that they can easily be overridden individually - 

e.g. 
"onMarkupRendered.logout": {
                    "this":   "{that}.dom.logout",
                    "method": "click",
                    "args":   "{that}.submitForm"
                }

incidentally - why does a click on the logout control submit the form?

the-t-in-rtf replied:
I split out the listeners as requested.

The "controls" component only has one thing it does itself, which is to make an AJAX call to GET /api/user/logout and then update the user in the model (setting it to null). That action is launched using the "submit" invoker of the templateFormControl grade.  This is the general job of the templateFormControl grade, to:

1. Render itself based on the initial model.
2. Submit an AJAX request when its submit invoker is called.
3. On success:
  a. Display whatever messages are configured using a particular rule.
  b. Update the module if desired based on another rule.
4. On error:
  a. Display whatever error messages are configured using a particular rule.

The templateFormControl grade replaced a bunch of forms, which is where it gots its name.  It is very similar to ajaxCapable, but adds standard wiring for success and error messages, and standard wiring for handling a "submit" action.

the-t-in-rtf replied:
I did note that there were no client side logout tests, and added those.

amb26 commented on src/js/client/login.js:
```diff
@@ -0,0 +1,103 @@
+                        message: "{login}.model.successMessage"
+                    },
+                    modelListeners: {
+                        // TODO:  Review with Antranig to confirm why the rules in `templateMessage` aren't enough to handle this.
```
Not clear what this comment is referring to - perhaps the base grade "templateFormControl"? I don't see that it has any modelListeners defined in it.

the-t-in-rtf replied:
I removed much of the unique material in these routers in favor of newer grades coming from `gpii-express` and `gpii-json-schema`.  The blocks that puzzled me earlier are now gone.

amb26 commented on src/js/server/api.js:
```diff
@@ -0,0 +1,133 @@
+fluid.defaults("gpii.express.user.api.hasMiddleware", {
+    gradeNames: ["gpii.express.user.api"],
+    components: {
+        session: {
```
Rather odd - why does this duplicate the middleware definition in "api"?

the-t-in-rtf replied:
It's a convenience grade to wire in the required session middleware if you are using this in an express instance (or under a router) that doesn't have it already.

amb26 commented on src/js/server/docs.js:
```diff
@@ -0,0 +1,32 @@
+// Display API docs written in Markdown
+"use strict";
+var fluid  = fluid || require("infusion");
```
Client-style

the-t-in-rtf replied:
Fixed.

amb26 commented on src/js/server/docs.js:
```diff
@@ -0,0 +1,32 @@
+var fs     = require("fs");
+var path   = require("path");
+
+var mdFile = path.resolve(__dirname, "../../docs/api.md");
```
Can now use module interpolation vars

the-t-in-rtf replied:
Done.

amb26 commented on src/js/server/forgot.js:
```diff
@@ -0,0 +1,195 @@
+                },
+                listeners: {
+                    "onRead.checkUser": {
+                        priority: "last",
```
What does this need to beat?

the-t-in-rtf replied:
```
 ____            _           _             
| __ ) _   _ ___| |_ ___  __| |            
|  _ \| | | / __| __/ _ \/ _` |            
| |_) | |_| \__ \ ||  __/ (_| |  _   _   _ 
|____/ \__,_|___/\__\___|\__,_| (_) (_) (_)
```

I think I had to do it in some unknown context and then held onto it as some kind of slavish cargo cult thing that had to be done for things to work.  I removed all unneeded priorities throughout and all tests still work.

amb26 replied:
https://www.youtube.com/watch?v=eIl8c-q0LII

amb26 replied:
I'm so BLOODY NORMAL
Yet I'm one of NATURE'S FREAKS!

amb26 commented on src/js/server/lib/hasMailerHandler.js:
```diff
@@ -0,0 +1,97 @@
+
+require("gpii-express");
+
+fluid.registerNamespace("gpii.express.user.api.hasMailHandler");
```
What is the grammatical usage of "has" in this case?

the-t-in-rtf replied:
"possesses".  This is a grade which is used by the "forgot" and "verify" endpoints.  Things that do not need or have a mail handler use the base grade rather than this one.

amb26 replied:
I think the modern tendency is to move towards the use of nouns rather than verbs in entity names - how about something like "withMailHandler" or "mailHandlerMixin"?

the-t-in-rtf replied:
I standardized both the filename and the component name to `withMailHandler`.

amb26 commented on src/js/server/lib/mailer.js:
```diff
@@ -0,0 +1,156 @@
+  // TODO:  Evaluate moving this to its own package, perhaps combining with gpii-mail-test.
+ */
+"use strict";
+var fluid = fluid || require("infusion");
```
client-style - fix globally

the-t-in-rtf replied:
Done.

amb26 commented on src/js/server/lib/mailer.js:
```diff
@@ -0,0 +1,156 @@
+
+require("gpii-handlebars");
+
+fluid.registerNamespace("gpii.express.user.mailer");
```
Why is this in the gpii.express namespace? I don't see any visible dependence on express infrastructure

the-t-in-rtf replied:
Even though it does not rely on express, it is part of this specific `gpii.express.user` package.  I could see evolving a `gpii-mailer` package, moving most of this code out, and keeping only the configuration here.

amb26 commented on src/js/server/lib/mailer.js:
```diff
@@ -0,0 +1,156 @@
+// output.
+//
+gpii.express.user.mailer.handlebars.sendTemplateMessage = function (that, mailOptions, context) {
+    var fullMailOptions = mailOptions ? mailOptions : {};
```
mailOptions || {} is more idiomatic

amb26 commented on src/js/server/lib/password.js:
```diff
@@ -0,0 +1,34 @@
+    A static function to consistently encode a password.  Used when storing passwords in CouchDB, and when logging in
+    using a username and password.
+
+    This package uses the `pbkdf2` functions provided by the `crypto` package built into node.  It has been tested to
```
Personally I would have preferred bcrypt but it seems that pbkdf2 is a perfectly reasonable choice: 
http://security.stackexchange.com/questions/4781/do-any-security-experts-recommend-bcrypt-for-password-storage

the-t-in-rtf replied:
`pbkdf2` was used because it provided complete backward compatibility with the specific couch-native legacy accounts we had already.  This avoided asking everyone to reset their passwords.  We could certainly add support for other schemes later if desired.

amb26 commented on src/js/server/lib/datasource.js:
```diff
@@ -0,0 +1,160 @@
+// A rewritten version of `kettle.dataSource.CouchDB.read` that allows rewriting of the response using the
+// transformation rules found in `options.rules.read`.  We cannot directly extend the existing couch grade in Kettle
+// because it has a non-namespaced listener for `onRead`, which we cannot replace. The `gradeNames` and `listener` block
```
This has been fixed in upcoming trunk Kettle - https://github.com/amb26/kettle/blob/KETTLE-32/lib/dataSource.js#L131

the-t-in-rtf replied:
Just so I understand, as this is a newer mechanism, how would I override a namespaced entry like this?

amb26 replied:
A listener to an event will displace another one on the same event with the same namespace. If you write a definition just like this one in a derived grade or in an options distribution, it will override this one. This particular mechanism isn't new, it has been implemented like this from the beginnings of the framework - although I remember that you and I established in a chat last year that it was, in the end, based on a "creative misunderstanding" of the jQuery system which it was believed to be modelled on. It turns out that jQuery namespaces don't behave like this at all - they are used at the "firing" end to select just a subset of the available listeners for a particular firing of an event. All of our listeners that are registered always fire all the time.

the-t-in-rtf replied:
Thanks.  Last question, is `"event.namespace"` as a listener key equivalent to `"event": { namespace: "namespace"}`?  i.e. will the shorter notation override the longer?

the-t-in-rtf replied:
We discussed this at the Eythorne summit of sixteen.  The two notations are equivalent.

amb26 commented on src/js/server/login.js:
```diff
@@ -0,0 +1,145 @@
+fluid.registerNamespace("gpii.express.user.api.login");
+
+gpii.express.user.api.login.post.handler.verifyPassword = function (that, response) {
+    if (that.request.body && that.request.body.username && that.request.body.password && response.salt && response.derived_key) {
```
Looks like a case for your validator - in case this check fails one imagines one might want to send an error

the-t-in-rtf replied:
I thought the same thing and prematurely removed that block before I remembered that this is actually checking the _CouchDB_ response to confirm whether the user exists.  I cleaned it up so that it's clearer what's happening.

amb26 commented on src/js/server/login.js:
```diff
@@ -0,0 +1,145 @@
+        }
+    }
+
+    that.sendResponse(401, { ok: false, message: "Invalid username or password."});
```
Ah, I see that you have been using the "early return" pattern - an "else" here would be logically clearer

the-t-in-rtf replied:
Without the early return I needed two else statements that sent the same response, but it's definitely clearer now.

amb26 commented on src/js/server/reset.js:
```diff
@@ -0,0 +1,183 @@
+// TODO: Check this against a schema using a schemaHandler
+gpii.express.user.api.reset.handler.checkResetCode = function (that, dataSourceResponse) {
+    // Prepare dates that will be used in later sanity checks.
+    var earliestAcceptable = new Date((new Date()).getTime() - that.options.codeExpiration);
```
Date.now() is clearer than this locution
https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Date/now

amb26 commented on tests/js/browser/browser-forgot-test.js:
```diff
@@ -0,0 +1,386 @@
+    var MailParser = require("mailparser").MailParser,
+        mailparser = new MailParser();
+
+    // If this ends up going any deeper, we will need to refactor using a testEnvironment and testCases.
```
It is already about 4 levels too deep : P

amb26 commented on tests/js/browser/browser-forgot-test.js:
```diff
@@ -0,0 +1,386 @@
+};
+
+gpii.express.user.tests.forgot.client.continueResetFromEmail = function (harness) {
+    var timestamp = (new Date()).getTime();
```
Date.now() - fix globally

amb26 commented on src/js/server/lib/loginRequired.js:
```diff
@@ -0,0 +1,61 @@
+                schemaUrl:  "http://ul.gpii.net/api/schemas/message.json",
+                invokers: {
+                    handleRequest: {
+                        // TODO:  Why can't we just use `{that}.sendResponse` here?
```
What happens if you do?

the-t-in-rtf replied:
Doesn't matter, the oddities here have been refactored away in favor of newer bits from `gpii-express` and `gpii-json-schema`.

amb26 commented on bower.json:
```diff
@@ -0,0 +1,31 @@
+    "Tony <tony@raisingthefloor.org>"
+  ],
+  "description": "A series of Fluid components that provide user management API and front-end components to interact with the API.",
+  "license": "See LICENSE file for details.",
```
Old-style bower file

the-t-in-rtf replied:
Fixed.

amb26 commented on src/js/client/resend.js:
```diff
@@ -0,0 +1,42 @@
+        },
+        rules: {
+            modelToRequestPayload: {
+                "":    "notfound", // Required to clear out the default rules from `templateFormControl`
```
We should do this with a mergePolicy - are there any cases of a user genuinely wanting to merge two rulesets together?

the-t-in-rtf replied:
I have used it myself, but have also frequently had to disable the defaults.  I will prepare a ticket and repo to split out the responsible `ajaxCapable` grade from `gpii-handlebars` and we can fix it, test it, and review it in more depth there.

the-t-in-rtf replied:
That work will be handled as part of [GPII-1587](https://issues.gpii.net/browse/GPII-1587).  I have left copious TODO notes where there are currently overrides or inherited rules.

amb26 commented on src/js/server/lib/datasource.js:
```diff
@@ -0,0 +1,161 @@
+fluid.registerNamespace("gpii.express.user.couchdb.writable");
+
+// A replacement for the default couch "write" operation that removes the hard-coded `value` wrapper in favor of rules.
+gpii.express.user.couchdb.writable.transformModelToCouch = function (that, model, options) {
```
I assume this is a good candidate for being replaced by your upcoming asymmetric dataSource

the-t-in-rtf replied:
Yes, I would propose waiting to do so until the new code has been reviewed and working on a smaller follow-up PR for the conversion.

the-t-in-rtf replied:
I added a TODO in this file and in all places where the current grade is used in this package.

amb26 commented on src/js/server/lib/datasource.js:
```diff
@@ -0,0 +1,161 @@
+};
+
+fluid.defaults("gpii.express.user.couchdb.writable", {
+    // We cannot directly extend the existing couch grade in Kettle because it has a non-namespaced listener for
```
Does so - https://github.com/amb26/kettle/blob/KETTLE-32/lib/dataSource.js#L482

the-t-in-rtf replied:
Yes, with the latest PR we can now use the Kettle writable couch grade, and I will do so in the "asymmetric" dataSource.  I will leave a TODO comment linking to that JIRA ticket.

amb26 commented on tests/js/server/password-function-tests.js:
```diff
@@ -0,0 +1,84 @@
+        jqUnit.assertEquals("The salt string should be of twice the byte length...", 48, saltString.length);
+
+        var saltInt = parseInt(saltString, 16);
+        jqUnit.assertNotEquals("The salt string should evaluate to a number when parsed as hex...", NaN, saltInt);
```
I'm not convinced this test tests anything since every number compares unequal to NaN. You can try https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger which is supported on node since before 0.12

amb26 replied:
https://kangax.github.io/compat-table/es6/

the-t-in-rtf replied:
Good catch, and fixed.

amb26 commented:
Test failure on "forgot" tests as follows:

15:44:52.937:  Sending a POST request to: http://localhost:8779/api/user/reset/efb147e1cdd00b3f95e209ab62578042 on port 8779
15:44:52.985:  DataSource Issuing HTTP request with options {
    "port": "8765",
    "method": "GET",
    "headers": {

    },
    "protocol": "http:",
    "auth": null,
    "host": "localhost:8765",
    "hostname": "localhost",
    "path": "/users/_design/lookup/_view/byResetCode?key=%22efb147e1cdd00b3f95e209ab62578042%22",
    "url": "http://localhost:8765/users/_design/lookup/_view/byResetCode?key=\"%code\"",
    "termMap": {
        "code": "%code"
    },
    "directModel": {
        "code": "efb147e1cdd00b3f95e209ab62578042"
    },
    "operation": "get",
    "reverse": false,
    "writeMethod": "PUT"
}
15:44:53.041:  Sending a POST request to: http://localhost:8779/api/user/login on port 8779
15:44:53.086:  DataSource Issuing HTTP request with options {
    "port": "8765",
    "method": "GET",
    "headers": {

    },
    "protocol": "http:",
    "auth": null,
    "host": "localhost:8765",
    "hostname": "localhost",
    "path": "/users/_design/lookup/_view/byUsernameOrEmail?key=%22existing%22",
    "url": "http://localhost:8765/users/_design/lookup/_view/byUsernameOrEmail?key=\"%username\"",
    "termMap": {
        "username": "%username"
    },
    "directModel": {
        "username": "existing",
        "password": "Password1"
    },
    "operation": "get",
    "reverse": false,
    "writeMethod": "PUT"
}
express deprecated res.send(status): Use res.sendStatus(status) instead node_modules\gpii-json-schema\node_modules\gpii-express\src\js\handler.js:77:38
15:44:53.135:  FATAL ERROR: Uncaught exception: Unexpected token O
SyntaxError: Unexpected token O
    at Object.parse (native)
    at gpii.express.user.api.reset.test.caseHolder.verifyResponse (E:\source\gits\gpii-express-user\tests\js\server\forgot-tests.js:162:48)
    at togo (E:\source\gits\gpii-express-user\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:248:33)
    at wrapped (E:\source\gits\gpii-express-user\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:323:26)
    at that.fire (E:\source\gits\gpii-express-user\node_modules\infusion\src\framework\core\js\Fluid.js:1399:40)
    at IncomingMessage.<anonymous> (E:\source\gits\gpii-express-user\node_modules\kettle\lib\test\KettleTestUtils.http.js:132:13)
    at emitNone (events.js:72:20)
    at IncomingMessage.emit (events.js:166:7)
    at endReadableNT (_stream_readable.js:903:12)
    at doNTCallback2 (node.js:439:9)
    at process._tickCallback (node.js:353:17)
15:44:53.139:  Current activity:
15:44:53.140:  jq: FAIL: Test name "Testing resetting a user's password, end-to-end..." - Message: Unexpected failure in test case (see following log for more d
etails): Unexpected token O
15:44:53.141:  jq: Source:     at kettle.test.handleUncaughtException (E:\source\gits\gpii-express-user\node_modules\kettle\lib\test\KettleTestUtils.js:31:15)
    at Object.that.fire (E:\source\gits\gpii-express-user\node_modules\infusion\src\framework\core\js\Fluid.js:1399:40)
    at process.onUncaughtException (E:\source\gits\gpii-express-user\node_modules\infusion\src\module\fluid.js:121:35)
    at emitOne (events.js:82:20)
    at process.emit (events.js:169:7)
    at process._fatalException (node.js:221:26)

Also 1 minor linting error

the-t-in-rtf commented:
This was another variation on the problems with updating the `sendResponse` method in `gpii.express.handler`.  Specifically it prevented `/hbs` from delivering template content, which killed the client-side components.  Turns out I had fixed only four of the five packages that depend on `gpii-express`, and not `gpii-pouchdb`.  I have fixed that and am testing now.

the-t-in-rtf commented:
That was it, my Windows build works again.  You'll have to pull the latest changes, and you may need to run `npm cache clear`, as the name of the package changed from `gpii-pouch` to the correct `gpii-pouchdb`.

amb26 commented:
These tests seem to run better than before, but I notice that they now have the behaviour of repeatedly creating and destroying a top-level electron window on the desktop. This is survivable but would be highly desirable to avoid. I do get two tests failures on the run though, here is the trail of the transcript:

14:32:54.795:  jq: FAIL: Test name "Create and verify a new user from end to end..." - Message: A login success message should now be displayed... - at sequence
 position 18 of 23
14:32:54.798:  jq: Source:     at pok (E:\source\gits\gpii-express-user\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:110:15)
    at jsUnitCompat.assertTrue (E:\source\gits\gpii-express-user\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:143:13)
    at togo (E:\source\gits\gpii-express-user\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:248:33)
    at wrapped (E:\source\gits\gpii-express-user\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:323:26)
    at Object.that.fire (E:\source\gits\gpii-express-user\node_modules\infusion\src\framework\core\js\Fluid.js:1399:40)
    at Nightmare.<anonymous> (E:\source\gits\gpii-express-user\node_modules\gpii-test-browser\src\js\nightmare.js:123:36)
    at Nightmare.done (E:\source\gits\gpii-express-user\node_modules\gpii-test-browser\node_modules\nightmare\lib\nightmare.js:225:15)
    at Nightmare.next (E:\source\gits\gpii-express-user\node_modules\gpii-test-browser\node_modules\nightmare\lib\nightmare.js:203:35)
    at EventEmitter.<anonymous> (E:\source\gits\gpii-express-user\node_modules\gpii-test-browser\node_modules\nightmare\lib\nightmare.js:213:12)
    at EventEmitter.g (events.js:260:16)
    at emitNone (events.js:67:13)
    at EventEmitter.emit (events.js:166:7)
    at ChildProcess.<anonymous> (E:\source\gits\gpii-express-user\node_modules\gpii-test-browser\node_modules\nightmare\lib\ipc.js:28:10)
    at emitTwo (events.js:87:13)
    at ChildProcess.emit (events.js:172:7)
    at handleMessage (internal/child_process.js:686:10)
    at Pipe.channel.onread (internal/child_process.js:440:11)
14:32:54.813:  Express stopped...
14:32:54.820:  Stopped mail server...
14:32:54.839:  jq: Test concluded - Test name "Create and verify a new user from end to end...": 1/2 passed - FAIL
14:32:54.848:  Deleted existing data from database 'users'...
14:32:54.851:  jq: ***************
14:32:54.854:  jq: All tests concluded: 103/105 total passed in 122324ms - FAIL
14:32:54.856:  jq: ***************

the-t-in-rtf commented:
I will look at the tests.  The creation of a "top level" Electron instance is a byproduct of using Nightmare, which previously was not obvious as we were using the defaults and hiding everything.  I'm not sure we have any control over that short of writing our own alternative to Nightmare, but I will look through their docs to confirm.

the-t-in-rtf commented:
Nightmare does not have any options that can be used to control the behavior of launching electron once for every browser instance.  Their own tests run with the same tell-tale twitch that demonstrates that electron is being launched and torn down for each run.

the-t-in-rtf commented:
On the other hand, I can avoid using the browser test environment for tests that don't actually require a browser.  That should reduce the twitching blackness that was part of the 78 "server" runs.

the-t-in-rtf commented:
The intermittent failures were the result of the tests running a few milliseconds before the page completed its initial render.  For now, adding a browser wait after the initial page load avoids the problems.  A better long-term fix is to get [GPII-1574](https://issues.gpii.net/browse/GPII-1574) working and wait until the client-side component finishes rendering before continuing the tests.
